modules/core/src/main/java/org/apache/ignite/internal/processors/rest/client/message/GridClientAuthenticationRequest.java

```diff
@@ -29,34 +32,55 @@
     private static final long serialVersionUID = 0L;
 
     /** Credentials. */
-    private Object cred;
+    private SecurityCredentials cred;
+
+    /** User attributes. */
+    private Map<String, String> userAttrs;
 
     /**
      * @return Credentials object.
      */
-    public Object credentials() {
+    public SecurityCredentials credentials() {
         return cred;
     }
 
     /**
      * @param cred Credentials object.
      */
-    public void credentials(Object cred) {
+    public void credentials(SecurityCredentials cred) {
         this.cred = cred;
     }
 
+    /**
+     * @return User attributes.
+     */
+    public Map<String, String> userAttributes() {
+        return userAttrs;
+    }
+
+    /**
+     * @param userAttrs User attributes.
+     */
+    public void userAttributes(Map<String, String> userAttrs) {
+        this.userAttrs = userAttrs;
+    }
+
     /** {@inheritDoc} */
     @Override public void writeExternal(ObjectOutput out) throws IOException {
         super.writeExternal(out);
 
         out.writeObject(cred);
+
+        U.writeMap(out, userAttrs);
     }
 
     /** {@inheritDoc} */
     @Override public void readExternal(ObjectInput in) throws IOException, ClassNotFoundException {
         super.readExternal(in);
 
-        cred = in.readObject();
+        cred = (SecurityCredentials)in.readObject();
+
+        userAttrs = U.readMap(in);
```

Comment:
Is this change compatible? Can we get this request from the "old" node in a rolling upgrade scenario?

Reply:
I don't know if it's compatible. I thought you have added tests to check that?